tools/prost-build/src/main.rs

```diff
@@ -0,0 +1,19 @@
+// Copyright 2020 Shift Cryptos AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+fn main() {
+    let messages_dir = std::env::var("MESSAGES_DIR").expect("MESSAGES_DIR env var not set");
```

Comment:
Why not use a command line flag?

Reply:
My issue with env vars is I see this:

```
    MESSAGES_DIR=${CMAKE_CURRENT_SOURCE_DIR}
    cargo run --manifest-path=${CMAKE_SOURCE_DIR}/tools/prost-build/Cargo.toml
```

and I have no idea whether it was a mistake, typo or `MESSAGES_DIR` is really used in the next command.

Reply:
idk, seemed consistent with OUT_DIR imposed by prost-build. And no CLI parsing needed ;)

I will look into it.

Reply:
yeah, that's even more confusing imho:

```rust
prost_build::compile_protos(&["hww.proto"], &[&messages_dir]).unwrap();
```

So, `compile_protos` takes in what to compile in function arguments, but somehow it wants an env var for the output... that looks very inconsistent, wouldn't you agree?

Reply:
i think OUT_DIR is imposed by the rust build.rs architecture :shrug:

Reply:
What about this? https://github.com/danburkert/prost/blob/8025627/prost-build/src/lib.rs#L476-L480

Reply:
good find! will try

Reply:
OUT_DIR is generally the right place to put stuff that you generate in a build.rs step. However if you want faster build times you might want cmake to do this step somehow. A bit like how we run bindgen and cbindgen. Then tool is only compiled once during the docker image build. Could also bring down the needed deps.

Not sure if this applies here 100% though.

Reply:
@x1ddos done

@NickeZ unfortunately build.rs cannot be used here due to the cargo bug linked in the commit msg.

I could still generate the file in the build directory and not in the src directory, but I always feel a bit uneasy about not committing generated protobuf files. The reasoning I put down was:

> - Less reliance on the tool to work when you check out and old commit
> - The rust crates do not depend on CMake in this way, which should
> make it easier to use standard Rust tooling (though more work is  - needed there).

What do you think?

Reply:
@NickeZ re docker: generally agree, but I would do it in a separate PR.

Reply:
I'm just saying that the reason that prost-build uses `OUT_DIR` is because cargo sets that environment variable when it runs a `build.rs` file. You should, on the other hand, use command line arguments to pass input files and output dirs.

I think `build.rs` files are useful for rust-only projects, but they are kind of evil in that they can do arbitrary things at compile time (like fetching stuff from ze internets). So I try to avoid them. In your case it is better to create a separate tool (like you've done) and run it separately from cargo. This reduces the number of deps on your main project and (when you move it to docker) reduces the overall build time.

Reply:
I'm split regarding if the generated sources should be committed. One benefit of committing them is that you notice if you by mistake use a tool that generates new/different output. (Maybe you locally have the wrong version installed for some reason..)

Reply:
Like @benma, I'd prefer to check in generated sources. We already do it in py/bitbox02 here and in api-go, api-js. Also, similar with vendoring Go pkgs in the wallet app. It's always nice to rely on git and see what's changed.

Reply:
> @x1ddos done

hm.. I still see no args parsing in main.rs. am I looking at the old code or wrong place?

Reply:
Hmm somehow I didn't push properly. Pushed again now.